### ISSUE TITLE
Unable to play videos on Mastodon in Safari

### ISSUE THREAD

**Issue by Casuallynoted** [tootsuite/mastodon#15872]
Safari version: Version 14.1 (16611.1.18.1.2)

I'm noticing that videos do not start playing via Safari on Mac. This also seems to affect mobile safari and any Mastodon iOS apps that I've used. Videos do however play on Firefox, Edge and Chrome on the same Mac. The errors I'm seeing in my console are:

Unhandled Promise Rejection: NotSupportedError: The operation is not supported.
anonymousfunction
rejectpromise

**Comment by Mennaruuk** [comment]
Hi! Do you mind dropping down a link to a sample video? I wanna see if I can reproduce the issue at least on my iPhone X.

**Comment by Casuallynoted** [comment]
> Hi! Do you mind dropping down a link to a sample video? I wanna see if I can reproduce the issue at least on my iPhone X.

Absolutely! This post from me was to test the issue after noticing it on friend's timelines:
https://casually.cat/@Casuallynoted/105861663232957272

**Comment by Mennaruuk** [comment]
I can definitely reproduce inability to play the sample video on iOS 14.4.1. Do you happen to know what version of Mastodon the instance is running? Perhaps it’s an outdated one.

**Comment by Casuallynoted** [comment]
> I can definitely reproduce inability to play the sample video on iOS 14.4.1. Do you happen to know what version of Mastodon the instance is running? Perhaps it’s an outdated one.

The instance is currently running on the latest release, 3.30. I just double checked to make sure it was the release build too :)

**Comment by Mennaruuk** [comment]
I’ll tag @Gargron for this, I think he can debug what’s going on. The videos I’ve seen on other instances do play ([example](https://mastodon.online/web/statuses/105855972327934431)), so I’m inclined to think there is an issue with the instance’s configuration. But I’ll defer to the experts, hold tight!

**Comment by Casuallynoted** [comment]
That's really bizarre, that does play for me. Maybe it has to do with the video codec?

**Comment by Casuallynoted** [comment]
Curious if there are any updates on this. I've noticed some GIFs don't seem to play either.

**Comment by daprice** [comment]
I can’t play videos or see gifs on a 4.0.0rc1 server (mastodon.online) or a 3.5.3 server (mastodon.gamedev.place) right now either. If I try to load the mp4 file directly, it doesn’t play that way either, just leaves "Failed to load resource: Plug-in handled load” in Safari's console. Loading the same mp4 in Firefox works fine.

Searching that error on StackOverflow brings up this: https://stackoverflow.com/questions/18103103/failed-to-load-resource-plugin-handled-load-on-ios suggesting that the video either needs to be served differently, or encoded with a “moov atom”.

**Comment by sznowicki** [comment]
Could you post some other example in mastodon context. This file works just fine on me on Safari/macOS 13.

Usually when it comes to videos Safari on previous OS (< iOS 16, < MacOS 12) it would fail because of wrong codec (e.g. webm). Now for latest versions Safari should support all popular codecs, including webm).

**Comment by daprice** [comment]
Hmm, after trying some different things to attempt to diagnose this, it turns out that simply restarting Safari fixed it (for now?). Whatever caused it must be a Safari or Webkit bug, not a Mastodon one.

**Comment by ansiwen** [comment]
I have a similar problem: both the official iOS App and Metatext can't play videos, but opening and playing in the browsers (Safari, Chrome, Firefox) works perfectly. (iOS: 16.1)

Example: https://chaos.social/@davedarko/109333133729107046

![image](https://user-images.githubusercontent.com/7792669/201533544-b1df6a4f-538a-430e-881d-412115ca7630.png)

After clicking on it, it shows:
![image](https://user-images.githubusercontent.com/7792669/201533628-7036fbcc-8311-4709-a159-47fc03d46bd7.png)

**Comment by ansiwen** [comment]
Ok, I found out what _my_ issue was: I run mastodon in a container, and it seems that the videos can't be played by iOS Apps, if they are served by the container/rails/puma. My guess would be some mime-type issue or something, or an issue with the reverse proxy settings. If the files under the `/system/` path are served directly by nginx, everything works perfectly.

**Comment by saschafoerster** [comment]
@ansiwen Could you post this part of your NGINX config maybe?

**Comment by ansiwen** [comment]
> @ansiwen Could you post this part of your NGINX config maybe?

@saschafoerster I bind-mounted a directory /x/y/z/system to /mastodon/public/system in the container, and then my config was something like
```
server {
...
    root /x/y/z/;
...
    location ~ ^/system/ {
        add_header Cache-Control "public, max-age=2419200, immutable";
        add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
        try_files $uri =404;
    }
... 
```

**but** I did some more research, and found out more why this is an issue in the first place:

The iOS library _AppleCoreMedia_ - which seems to be used by Apps to play videos - (and  probably also older Safaris) uses and requires byte-range requests, as mentioned [here](https://developer.apple.com/library/archive/documentation/AppleApplications/Reference/SafariWebContent/CreatingVideoforSafarioniPhone/CreatingVideoforSafarioniPhone.html#//apple_ref/doc/uid/TP40006514-SW6). It is standard compliant to respond to byte-range requests with 200 and the complete file, the library ignores them and only accept 206 responses with the requested range.

The cache of the reverse proxy of Nginx supports byte-range requests, but honours them _only_ if the "Accept-Ranges" is set by the client, which - surprise, surprise - _AppleCoreMedia_ does not do. 🙄 Fortunately there is an nginx configuration option [proxy_force_ranges](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_force_ranges) which makes nginx always respond with 206 if a range has been requested. So my proxy cache config now looks like this:

```
  location @proxy {
...
    proxy_pass http://backend;
    proxy_buffering on;
    proxy_redirect off;
    proxy_http_version 1.1;
    proxy_force_ranges on;    
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;
...
  }
```

and everything works perfectly!

I hope other people having the same problem will find this here and I can save them the half day of research. 😅

**Comment by saschafoerster** [comment]
Worked for me!

**Comment by Happyfeet01** [comment]
Worked for me. Big Thanks!

**Comment by ramuuns** [comment]
For anyone else who runs into this, and has additionally an s3 (with your own proxy in front of it), you also need the 
```
proxy_force_ranges on; 
```
in the nginx configuration of your s3 proxy as well.

**Comment by zeroesca** [comment]
> For anyone else who runs into this, and has additionally an s3 (with your own proxy in front of it), you also need the
> 
> ```
> proxy_force_ranges on; 
> ```
> 
> in the nginx configuration of your s3 proxy as well.

👌 -- thanks!
